src/txrequest.cpp

```diff
@@ -0,0 +1,238 @@
+        it = Erase<ByTxid>(it);
+    }
+}
+
+static const uint256 UINT256_ZERO;
+
+void TxRequestTracker::DeletedPeer(uint64_t peer)
+{
+    auto& index = m_index.get<ByPeer>();
+    auto it = index.lower_bound(EntryPeer{peer, false, UINT256_ZERO});
+    while (it != index.end() && it->m_peer == peer) {
+        // Check what to continue with after this iteration. Note that 'it' may change position, and std::next(it)
+        // may be deleted in the process, so this needs to be decided beforehand.
+        auto it_next = (std::next(it) == index.end() || std::next(it)->m_peer != peer) ? index.end() : std::next(it);
+        // If the entry isn't already COMPLETED, first make it COMPLETED (which will mark other CANDIDATEs as
+        // CANDIDATE_BEST, or delete all of a txid's entries if no non-COMPLETED ones are left).
+        if (MakeCompleted(m_index.project<ByTxid>(it))) {
+            // Then actually delete the entry (unless it was already deleted by MakeCompleted).
+            Erase<ByPeer>(it);
+        }
+        it = it_next;
+    }
+}
+
+void TxRequestTracker::ReceivedInv(uint64_t peer, const uint256& txid, std::chrono::microseconds reqtime)
+{
+    // Bail out if we already have a CANDIDATE_BEST entry for this (txid, peer) combination. The case where there is
+    // a non-CANDIDATE_BEST entry already will be caught by the uniqueness property of the ByPeer index
+    // automatically.
+    if (m_index.get<ByPeer>().count(EntryPeer{peer, true, txid})) return;
+
+    auto ret = m_index.get<ByPeer>().emplace(txid, peer, reqtime, m_sequence);
+    if (ret.second) {
+        auto it = m_index.project<ByTxid>(ret.first);
+        ++m_peerinfo[peer].m_total;
+        ++m_sequence;
+        if ((it == m_index.get<ByTxid>().begin() || std::prev(it)->m_txid != txid) &&
+            (std::next(it) == m_index.get<ByTxid>().end() || std::next(it)->m_txid != txid)) {
+            // This is both the first and the last entry for a given txid; set its m_first.
+            ret.first->m_first = true;
+        }
+    }
+}
+
+void TxRequestTracker::RequestedTx(uint64_t peer, const uint256& txid, std::chrono::microseconds now)
+{
+    auto it = m_index.get<ByPeer>().find(EntryPeer{peer, true, txid});
+    // RequestedTx can only be called on CANDIDATE_BEST entries (this is implied by its condition that it can only be
+    // called on txids returned by GetRequestable (and only AlreadyHave and RequestedTx can be called in between,
+    // which preserve the state of other txids).
+    assert(it != m_index.get<ByPeer>().end());
+    assert(it->GetState() == State::CANDIDATE_BEST);
+    Modify<ByPeer>(it, [now,this](Entry& entry) {
+        entry.SetState(State::REQUESTED);
+        entry.m_time = now + m_timeout;
+    });
+}
+
+void TxRequestTracker::ReceivedResponse(uint64_t peer, const uint256& txid)
+{
+    // We need to search the ByPeer index for both (peer, false, txid) and (peer, true, txid).
```

Comment:
259fb92

The ByPeer index second member as true signals a state equal to CANDIDATE_BEST but I'm not sure if you can effectively hit here. An entry state is _BEST only between `GetRequestable` and either `RequestedTx/AlreadyHaveTx`, both of them respectively transitioning to REQUESTED/entry removal.

Reply:
As explained elsewhere, no, BEST exists for much longer than that, and I'm 99% sure this can be hit.